mwdb/resources/oauth.py

```diff
@@ -681,3 +691,75 @@ def get(self, provider_name):
 
         schema = OpenIDLogoutLinkResponseSchema()
         return schema.dump({"url": provider.logout_endpoint})
+
+
+@rate_limited_resource
+class OpenIDDiscoveryResource(Resource):
+    @requires_authorization
+    def post(self):
+        """
+        ---
+        summary: Get discovery endpoint data
+        description: |
+            Get data from OIDC discovery endpoint
+        security:
+            - bearerAuth: []
+        tags:
+            - auth
+        requestBody:
+            description: URL to discovery endpoint
+            content:
+              application/json:
+                schema: OpenIDDiscoveryURLRequestSchema
+        responses:
+            200:
+                description: When data was successfully downloaded
+                content:
+                  application/json:
+                    schema: OpenIDDiscoveryDataResponseSchema
+            400:
+                description: Error occurred
+            404:
+                description: Invalid URL
+            503:
+                description: |
+                    Request canceled due to connection timeout.
+        """
+        schema = OpenIDDiscoveryURLRequestSchema()
+        obj = loads_schema(request.get_data(as_text=True), schema)
+        url = obj.get("url")
+
+        if url is None:
+            raise NotFound("URL cannot be empty")
+        try:
+            response = requests.get(url)
+        except requests.exceptions.MissingSchema:
+            raise NotFound(
+                "Cannot connect to given URL. Remember to use http:// or https://"
+            )
+        except requests.exceptions.ConnectionError:
+            raise NotFound("Cannot connect to given URL")
+        except requests.exceptions.Timeout:
+            raise ServiceUnavailable("Request timed out")
+        except requests.exceptions.RequestException as e:
+            raise BadRequest(e)
```

Comment:
In a potential traceback it would be nice to see what those exceptions are related to. If they are not meant to be seen as "caused because of `request.exceptions.SomeError`", simply put a `from None` after every `raise ...`.

Currently, the traceback in case of a `request.exceptions.ConnectionError` error would be:
```
Traceback (most recent call last):
  File "...", line ..., in ...
    (a line that raises the request.exceptions.ConnectionError exception)
  request.exceptions.ConnectionError: ...

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "mwdb/resources/oauth.py", line 737, in post
    raise NotFound(
  NotFound: Cannot connect to given URL. Remember to use http:// or https://
```

But after a slight modification like this:

```py
try:
    response = requests.get(url)
except requests.exceptions.MissingSchema as exc:
    raise NotFound(
        "Cannot connect to given URL. Remember to use http:// or https://"
    ) from exc
```

the traceback seems much more like the one you would intend:

```
Traceback (most recent call last):
  File "...", line ..., in ...
    (a line that raises the request.exceptions.ConnectionError exception)
  request.exceptions.ConnectionError: ...

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "mwdb/resources/oauth.py", line 737, in post
    raise NotFound(
  NotFound: Cannot connect to given URL. Remember to use http:// or https://
```

See also https://docs.python.org/3/tutorial/errors.html#exception-chaining for the mentioned `from None` case.